storage/flash_common.c

```diff
@@ -115,13 +117,57 @@ secbool flash_area_write_word(const flash_area_t *area, uint32_t offset,
   return flash_write_word(sector, sector_offset, data);
 }
 
-secbool flash_area_write_quadword(const flash_area_t *area, uint32_t offset,
-                                  const uint32_t *data) {
+secbool flash_area_write_block(const flash_area_t *area, uint32_t offset,
+                               const flash_block_t block) {
+  if (!FLASH_IS_ALIGNED(offset)) {
+    return secfalse;
+  }
+
+  uint16_t sector;
+  uint32_t sector_offset;
+  if (sectrue != get_sector_and_offset(area, offset, &sector, &sector_offset)) {
+    return secfalse;
+  }
+
+#if FLASH_BLOCK_WORDS == 1
+  return flash_write_word(sector, sector_offset, block);
+#else
+
+  uint32_t *addr =
+      (uint32_t *)flash_area_get_address(area, offset, 4 * sizeof(uint32_t));
+
+  if (addr == NULL) {
+    return secfalse;
+  }
+
+  secbool old_all_ff = sectrue;
+  secbool new_all_00 = sectrue;
+  secbool all_equal = sectrue;
+
   for (int i = 0; i < 4; i++) {
-    if (sectrue !=
-        flash_area_write_word(area, offset + i * sizeof(uint32_t), data[i])) {
+    if (addr[i] != 0xFFFFFFFF) {
+      old_all_ff = secfalse;
+    }
+    if (block[i] != 0x00000000) {
+      new_all_00 = secfalse;
+    }
+    if (addr[i] != ((uint32_t *)block)[i]) {
+      all_equal = secfalse;
+    }
+  }
+
+  if (!(old_all_ff == sectrue || new_all_00 == sectrue ||
+        all_equal == sectrue)) {
+    return secfalse;
+  }
```

Comment:
Why do we need this? Shouldn't the underlying implementation of `flash_write_word()` be doing this internally? Also, if `FLASH_BIT_ACCESS` is true, then this is too restrictive. (I think `FLASH_BIT_ACCESS` and `FLASH_BLOCK_WORDS` should be treated as independent.)

Reply:
this will be part of U5 flash driver implementation (specifically its `flash_write_quadword` function) when we have one (its part of https://github.com/trezor/trezor-firmware/pull/3370 ), but i also need something here to make tests work. Probably part of `tests/c/flash.c` is better place though? (compiled conditionally depending on the `FLASH_BIT_ACCESS` and `FLASH_BLOCK_WORDS` )

Reply:
In that case it definitely belongs into tests/c/flash.c.

Reply:
moved https://github.com/trezor/trezor-firmware/commit/31ca162522180b08acda83e0f00f8e615f1e8999